receiver/awscontainerinsightreceiver/internal/gpu/dcgmscraper.go

```diff
@@ -0,0 +1,232 @@
+				Regex:        relabel.MustNewRegexp("([^:]+)(?::\\d+)?"),
+				Replacement:  "${1}:9400",
+				TargetLabel:  "__address__",
+				Action:       relabel.Replace,
+			},
+		},
+		MetricRelabelConfigs: []*relabel.Config{
+			{
+				SourceLabels: model.LabelNames{"__name__"},
+				Regex:        relabel.MustNewRegexp("DCGM_.*"),
+				Action:       relabel.Keep,
+			},
+			{
+				SourceLabels: model.LabelNames{"Hostname"},
+				TargetLabel:  "NodeName",
+				Regex:        relabel.MustNewRegexp("(.*)"),
+				Replacement:  "${1}",
+				Action:       relabel.Replace,
+			},
+			{
+				SourceLabels: model.LabelNames{"namespace"},
+				TargetLabel:  "Namespace",
+				Regex:        relabel.MustNewRegexp("(.*)"),
+				Replacement:  "${1}",
+				Action:       relabel.Replace,
+			},
+			// hacky way to inject static values (clusterName & instanceId) to label set without additional processor
+			// relabel looks up an existing label then creates another label with given key (TargetLabel) and value (static)
+			{
+				SourceLabels: model.LabelNames{"namespace"},
+				TargetLabel:  "ClusterName",
+				Regex:        relabel.MustNewRegexp(".*"),
+				Replacement:  opts.HostInfoProvider.GetClusterName(),
+				Action:       relabel.Replace,
+			},
+			{
+				SourceLabels: model.LabelNames{"namespace"},
+				TargetLabel:  "InstanceId",
+				Regex:        relabel.MustNewRegexp(".*"),
+				Replacement:  opts.HostInfoProvider.GetInstanceID(),
+				Action:       relabel.Replace,
+			},
+			{
+				SourceLabels: model.LabelNames{"pod"},
+				TargetLabel:  "FullPodName",
+				Regex:        relabel.MustNewRegexp("(.*)"),
+				Replacement:  "${1}",
+				Action:       relabel.Replace,
+			},
+			// additional k8s podname for service name decoration
+			{
+				SourceLabels: model.LabelNames{"pod"},
+				TargetLabel:  "K8sPodName",
+				Regex:        relabel.MustNewRegexp("(.*)"),
+				Replacement:  "${1}",
+				Action:       relabel.Replace,
+			},
+			{
+				SourceLabels: model.LabelNames{"container"},
+				TargetLabel:  "ContainerName",
+				Regex:        relabel.MustNewRegexp("(.*)"),
+				Replacement:  "${1}",
+				Action:       relabel.Replace,
+			},
+			{
+				SourceLabels: model.LabelNames{"device"},
+				TargetLabel:  "GpuDevice",
+				Regex:        relabel.MustNewRegexp("(.*)"),
+				Replacement:  "${1}",
+				Action:       relabel.Replace,
+			},
+		},
+	}
+
+	promConfig := prometheusreceiver.Config{
+		PrometheusConfig: &config.Config{
+			ScrapeConfigs: []*config.ScrapeConfig{scrapeConfig},
+		},
+	}
+
+	params := receiver.CreateSettings{
+		TelemetrySettings: opts.TelemetrySettings,
+	}
+
+	decoConsumer := decorateConsumer{
+		containerOrchestrator: ci.EKS,
+		nextConsumer:          opts.Consumer,
+		k8sDecorator:          opts.K8sDecorator,
+		logger:                opts.Logger,
+	}
+
+	promFactory := prometheusreceiver.NewFactory()
+	promReceiver, err := promFactory.CreateMetricsReceiver(opts.Ctx, params, &promConfig, &decoConsumer)
```

Comment:
Can you explain what will happen if there is no DCGM exporter running on the node?  How often does the scraper retry?  Will we get a bunch of errors in the logs?

Reply:
The scraper will print out en error message every minute if dcgm exporter is not available. If this becomes concern, we could probably check the instance type with`EnableGpuMetrics` flag to decide whether to register gpu scraper or not.

Reply:
Yes, it will be very confusing to see those errors if we enable the scraper on a non-GPU node.  Can the operator only set `enable_gpu_metrics` if it also deploys DCGM exporter on a node?  Or can the agent do something, like only start up the Prometheus scraper if it sees a local DCGM exporter?

Reply:
Sorry you're right, I should say add-on, not operator.

Can you show what the log will look like on a non-GPU node?

I will say this is a blocker, it's going to affect every non-GPU customer.  I like your idea of checking for DCGM exporter.  Or can we do something quick to swallow the error message?  (Although we don't want to lose it for troubleshooting purposes if there are real errors contacting DCGM exporter on a GPU node.)

Reply:
Actually, it will print a Debug log looks like:
```
 2024-03-01T14:06:52Z D! {"caller":"scrape/scrape.go:1351","msg":"Scrape failed","kind":"receiver","name":"awscontainerinsightreceiver","data_type":"metr │
│ ics","scrape_pool":"containerInsightsDCGMExporterScraper","target":"http://dcgm-exporter-service.amazon-cloudwatch.svc:9400/metrics","error":"Get \"http │
│ ://dcgm-exporter-service.amazon-cloudwatch.svc:9400/metrics\": context deadline exceeded"}
```
I think we can troubleshoot with running the agent in debug mode in case there is an issue. For other cases, scraper won't clutter the logs. 

As the long term solution, we should not run the scraper at all on non-GPU instances. I will start work on that DCGM pod scanning option.

Reply:
OK that's great, debug line is fine.  Agree on your plans for the next phase as well.